demo/src/main/java/com/example/demo/web/controllers/UserController.java

```diff
@@ -1,58 +1,143 @@
+
+        ForkJoinPool.commonPool().submit(() -> {
+            logger.info("Processing in separate thread");
+            List<User> list = userService.getAll();
+            output.setResult(list);
+        });
+
+        logger.info("Thread freed");
+        return output;
     }
-    // end::get-aggregate-root[]
+
 
     @GetMapping("/random")
-    User randomUser() {
-        return userService.getRandomUser();
-    }
+    DeferredResult<User> randomUser() {
+        logger.info("Received random user request");
+        DeferredResult<User> output = new DeferredResult<>(5L, new UserNotFoundException());
 
-    @PostMapping("/users")
-    User newUser(@RequestBody User newUser) {
-        return userService.addUser(newUser);
+        ForkJoinPool.commonPool().submit(() -> {
+            logger.info("Processing in separate thread");
+            User user = userService.getRandomUser();
+            output.setResult(user);
+        });
+
+        logger.info("Thread freed");
+        return output;
     }
 
-    // Single item
 
     @GetMapping("/users/{id}")
-    User one(@PathVariable Long id) {
-        return userService.getUserById(id);
+    DeferredResult<User> one(@PathVariable Long id) {
+        logger.info("Received get user by ID request");
+        DeferredResult<User> output = new DeferredResult<>(5L, new UserNotFoundException(id));
+
+        ForkJoinPool.commonPool().submit(() -> {
+            logger.info("Processing in separate thread");
+            User user = userService.getUserById(id);
+            output.setResult(user);
+        });
+
+        logger.info("Thread freed");
+        return output;
+    }
+
+    private void matchOpponents() {
+        if (this.usersReadyToPlay.size() < 2) {
+            return;
+        }
+
+        var entries = this.usersReadyToPlay.entrySet();
+
+        for (Iterator<Map.Entry<DeferredResult<User>, User>> it = entries.iterator(); it.hasNext(); ) {
+            var firstUser = it.next();
+            Map.Entry<DeferredResult<User>, User> secondUser;
+
+            if (!it.hasNext()) {
+                break;
+            } else {
+                secondUser = it.next();
+            }
+
+            firstUser.getKey().setResult(secondUser.getValue());
+            secondUser.getKey().setResult(firstUser.getValue());
+
+            // TODO: create new game with given users
+            userService.addUser(firstUser.getValue());
+            userService.addUser(secondUser.getValue());
+        }
+    }
+
+    @PostMapping("/users")
+    DeferredResult<User> newUser(@RequestBody User newUser) {
+        final DeferredResult<User> result = new DeferredResult<>(5L, new OpponentNotFoundException());
+
+        this.usersReadyToPlay.put(result, newUser);
+
+        result.onCompletion(() -> usersReadyToPlay.remove(result));
+
+        result.onTimeout(() -> usersReadyToPlay.remove(result));
+
+        matchOpponents();
```

Comment:
I don't think opponents matching logic is part of controller. 
It can be separate service like matchingService, which can be used by UserService or GameState service, depending on your vision.